File: scripts/ghidra_apply_matches.py
# Ghidra (Jython) post-script: apply pc_ps2_matches_full.csv.
#  conf>=0.9  -> real function name (PS2 name)
#  conf< 0.9  -> plate comment "PS2 candidate: <name> (conf x)" (draft tier)
# Args: <pc_ps2_matches_full.csv>
# @category GiantsRE
import csv
from ghidra.program.model.symbol import SourceType
from ghidra.app.cmd.function import CreateFunctionCmd
from ghidra.util.task import ConsoleTaskMonitor

ARGS = getScriptArgs()
CSV_PATH = ARGS[0]

fm = currentProgram.getFunctionManager()
af = currentProgram.getAddressFactory().getDefaultAddressSpace()
listing = currentProgram.getListing()
mon = ConsoleTaskMonitor()

renamed = commented = missed = 0
with open(CSV_PATH) as f:
    for row in csv.DictReader(f):
        va = int(row["pc_addr"], 16)
        name = row["ps2_name"]
        conf = float(row["conf"])
        addr = af.getAddress(va)
        fn = fm.getFunctionContaining(addr)
        if fn is None:
            cmd = CreateFunctionCmd(addr)
            cmd.applyTo(currentProgram, mon)
            fn = fm.getFunctionContaining(addr)
        if fn is None:
            missed += 1
            continue
        if conf >= 0.9:
            try:
                fn.setName(name, SourceType.USER_DEFINED)
                renamed += 1
                continue
            except Exception:
                pass
        plate = listing.setPlate(addr, "PS2 candidate: %s (conf %.2f)" % (name, conf))
        if plate:
            commented += 1
print("matches: renamed=%d candidate-commented=%d no-func=%d" % (renamed, commented, missed))

File: scripts/ghidra_apply_matches.java
// GiantsRE: apply pc_ps2_matches_full.csv (>=0.9 rename; else plate comment).
// Args: <pc_ps2_matches_full.csv>
// @category GiantsRE
import java.io.File;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;
import ghidra.app.script.GhidraScript;
import ghidra.app.cmd.function.CreateFunctionCmd;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.CodeUnit;
import ghidra.program.model.symbol.SourceType;
import ghidra.util.task.ConsoleTaskMonitor;

public class ghidra_apply_matches extends GhidraScript {
    public void run() throws Exception {
        String csvPath = getScriptArgs()[0];
        double minRename = getScriptArgs().length > 1 ? Double.parseDouble(getScriptArgs()[1]) : 0.9;
        String draftPrefix = getScriptArgs().length > 2 ? getScriptArgs()[2] : "ps2_";
        double minDraft = getScriptArgs().length > 3 ? Double.parseDouble(getScriptArgs()[3]) : 0.35;
        var fm = currentProgram.getFunctionManager();
        var af = currentProgram.getAddressFactory().getDefaultAddressSpace();
        var listing = currentProgram.getListing();
        int renamed = 0, commented = 0, missed = 0;
        Pattern p = Pattern.compile("^([0-9a-fA-F]+),[^,]*,[^,]*,([^,]*),([0-9.]+),");
        for (String line : Files.readAllLines(new File(csvPath).toPath())) {
            Matcher m = p.matcher(line);
            if (!m.find()) continue;
            long va = Long.parseLong(m.group(1), 16);
            String name = m.group(2).trim();
            double conf = Double.parseDouble(m.group(3));
            Address addr = af.getAddress(va);
            Function fn = fm.getFunctionContaining(addr);
            if (fn == null) {
                new CreateFunctionCmd(addr).applyTo(currentProgram, monitor);
                fn = fm.getFunctionContaining(addr);
            }
            if (fn == null) { missed++; continue; }
            if (conf >= minRename) {
                try { fn.setName(name, SourceType.USER_DEFINED); renamed++; continue; }
                catch (Exception ex) { /* fallthrough to comment */ }
            }
            { CodeUnit cu = listing.getCodeUnitAt(addr); if (cu != null) cu.setComment(CodeUnit.PLATE_COMMENT, "PS2 candidate: " + name + " (conf " + conf + ")"); }
                commented++;

        }
        println("matches: renamed=" + renamed + " candidate-commented=" + commented + " no-func=" + missed);
    }
}
